File: backend/src/services/claudeAIService.ts
import Anthropic from "@anthropic-ai/sdk";
import { getClaudeSettings } from "./claudeSettingsService";
import type { AiChatRequest, AiChatResult } from "./aiProviderService";

/**
 * The one service that talks to the Anthropic Claude Messages API. Every
 * caller goes through aiProviderService.ts's `sendMessage` (which dispatches
 * here only when Claude is the active provider) — nothing else in this
 * codebase constructs a Claude request directly.
 *
 * Unlike fabrixAIService.ts's callFabrixAgent, there's no unverified wire
 * contract to isolate here — this is a standard `client.messages.create`
 * call against the official, documented Messages API. Claude's messages
 * array only supports "user"/"assistant" roles (no "tool" role), so every
 * "system"-role AiChatTurn is merged into the top-level `system` string, and
 * "tool"-role turns (this app's TOOL RESULTS-labeled content, injected back
 * into the conversation — see aiAssistantService.ts) are sent as "user"
 * turns, same as how fabrixAIService.ts folds them into its own flat
 * contents array. The labeling the model relies on lives in the text itself
 * (section headers like "CAMPAIGN DATA"/"TOOL RESULTS"/"USER MESSAGE"), not
 * in the role field, so this collapsing is safe.
 */
export async function sendMessage(request: AiChatRequest): Promise<AiChatResult> {
  const settings = await getClaudeSettings();
  if (!settings) {
    return { ok: false, error: "Claude is not configured" };
  }
  if (!settings.enabled) {
    return { ok: false, error: "Claude is disabled" };
  }

  const client = new Anthropic({ apiKey: settings.apiKey });

  const systemPrompt = request.messages
    .filter((m) => m.role === "system")
    .map((m) => m.content)
    .join("\n\n");
  const messages: Anthropic.MessageParam[] = request.messages
    .filter((m) => m.role !== "system")
    .map((m) => ({ role: m.role === "assistant" ? "assistant" : "user", content: m.content }));

  const startedAt = Date.now();
  try {
    const response = await client.messages.create({
      model: settings.model,
      max_tokens: 8192,
      ...(systemPrompt ? { system: systemPrompt } : {}),
      messages,
    });

    if (response.stop_reason === "refusal") {
      console.error(`[claudeAIService] model=${settings.model} durationMs=${Date.now() - startedAt} status=refusal`);
      return { ok: false, error: "Claude declined to respond to this request." };
    }

    const textBlock = response.content.find((block): block is Anthropic.TextBlock => block.type === "text");
    if (!textBlock) {
      return { ok: false, error: "Claude response did not include any text" };
    }

    console.log(`[claudeAIService] model=${settings.model} durationMs=${Date.now() - startedAt} status=ok`);
    return {
      ok: true,
      replyText: textBlock.text,
      model: response.model,
      tokenUsage: response.usage.input_tokens + response.usage.output_tokens,
    };
  } catch (err) {
    const message = describeClaudeError(err);
    console.error(`[claudeAIService] model=${settings.model} durationMs=${Date.now() - startedAt} status=error error=${message}`);
    return { ok: false, error: message };
  }
}

/** Most-specific-first exception chain — never string-match a raw error
 * message, per the typed Anthropic.* exception classes the SDK exports. */
function describeClaudeError(err: unknown): string {
  if (err instanceof Anthropic.AuthenticationError) return "Claude authentication failed — check the API key.";
  if (err instanceof Anthropic.RateLimitError) return "Claude rate limit exceeded — try again shortly.";
  if (err instanceof Anthropic.BadRequestError) return `Claude rejected the request: ${err.message}`;
  if (err instanceof Anthropic.APIError) return `Claude error (${err.status}): ${err.message}`;
  return err instanceof Error ? err.message : String(err);
}

File: backend-py/app/services/claudeAIService.py
"""Port of `backend/src/services/claudeAIService.ts`.

Talks to the Anthropic Claude Messages API. Uses the official `anthropic` SDK.
"""

from __future__ import annotations

import time
from typing import Any, Optional

import anthropic

from app.services.claude_settings_service import ClaudeSettings, get_claude_settings


async def send_message(request: dict[str, Any], db: Any) -> dict[str, Any]:
    """Sends one conversation turn to Claude and returns its reply, or a
    structured ``{ok: False, error: ...}`` if the config is missing/disabled."""
    settings = get_claude_settings(db)
    if settings is None:
        return {"ok": False, "error": "Claude is not configured"}
    if not settings.enabled:
        return {"ok": False, "error": "Claude is disabled"}

    client = anthropic.Anthropic(api_key=settings.apiKey)

    system_parts = [m["content"] for m in request["messages"] if m["role"] == "system"]
    system_prompt = "\n\n".join(system_parts) if system_parts else ""

    messages = []
    for m in request["messages"]:
        if m["role"] == "system":
            continue
        role = "assistant" if m["role"] == "assistant" else "user"
        messages.append({"role": role, "content": m["content"]})

    started_at = time.time()
    try:
        kwargs: dict[str, Any] = {
            "model": settings.model,
            "max_tokens": 8192,
            "messages": messages,
        }
        if system_prompt:
            kwargs["system"] = system_prompt

        response = client.messages.create(**kwargs)

        if response.stop_reason == "refusal":
            elapsed_ms = (time.time() - started_at) * 1000
            print(f"[claudeAIService] model={settings.model} durationMs={elapsed_ms:.0f} status=refusal")
            return {"ok": False, "error": "Claude declined to respond to this request."}

        text_block = next(
            (block for block in response.content if block.type == "text"), None
        )
        if text_block is None:
            return {"ok": False, "error": "Claude response did not include any text"}

        elapsed_ms = (time.time() - started_at) * 1000
        print(f"[claudeAIService] model={settings.model} durationMs={elapsed_ms:.0f} status=ok")
        return {
            "ok": True,
            "replyText": text_block.text,
            "model": response.model,
            "tokenUsage": response.usage.input_tokens + response.usage.output_tokens,
        }
    except anthropic.AuthenticationError:
        return {"ok": False, "error": "Claude authentication failed — check the API key."}
    except anthropic.RateLimitError:
        return {"ok": False, "error": "Claude rate limit exceeded — try again shortly."}
    except anthropic.BadRequestError as exc:
        return {"ok": False, "error": f"Claude rejected the request: {exc.message}"}
    except anthropic.APIError as exc:
        return {"ok": False, "error": f"Claude error ({exc.status_code}): {exc.message}"}
    except Exception as exc:
        return {"ok": False, "error": str(exc)}
